### PR TITLE
[DEVOPS-74] ekg statsd backend

### PR DESCRIPTION
**Pull request by avieth**
`EkgParams`, `StatsdParams` options defined. Also the `metrics` flag which must be present in order for EKG and statsd backends to function. There's already a `stats` flag which seems to do something else entirely; should we use it for this as well?

### CODE REVIEW

**Comment by avieth** [comment]
@domenkozar

File: scripts/common.sh [hunk 1]
```diff
@@ -165,7 +166,9 @@ function node_cmd {
   echo -n " --kademlia-dump-path  $(dump_path $kademlia_dump_path)"
   echo -n " --system-start $system_start"
   echo -n " --peer-id $peer_id"
-  # echo -n " --monitor-port $monitor_port +RTS -T -RTS "
+  echo -n " --metrics +RTS -T -RTS"
```
Comment (dcoutts-2fa) [inline scripts/common.sh]: We can if we want build in the +RTS -T param at compile time. I assume that would be preferable.

File: src/Pos/Launcher/Runner.hs [hunk 2]
```diff
@@ -176,11 +181,34 @@ runRawRealMode peerId transport np@NodeParams {..} sscnp listeners outSpecs (Act
...
+                                    , Monitoring.prefix = statsdPrefix
+                                    , Monitoring.suffix = statsdSuffix
+                                    }
+                            liftIO $ Monitoring.forkStatsd statsdOptions ekgStore'
+                    return (mEkgServer, mStatsdServer)
+
+        let stopMonitoring Nothing = return ()
+            stopMonitoring (Just (mEkg, mStatsd)) = do
+                maybe (pure ()) (killThread . Monitoring.statsdThreadId) mStatsd
+                maybe (pure ()) stopMonitor mEkg
```
Comment (dcoutts-2fa) [inline src/Pos/Launcher/Runner.hs]: This section is getting rather long. Can we move it out of line, or at least into a where clause?

File: src/Pos/Launcher/Runner.hs [hunk 3]
```diff
@@ -262,8 +290,7 @@ runServer peerId mkTransport mkReceiveDelay packedLS_M (OutSpecs wouts) withNode
     logInfo $ sformat ("Our verInfo "%build) ourVerInfo
     node mkTransport mkReceiveDelay stdGen BiP (peerId, ourVerInfo) defaultNodeEnvironment $ \__node ->
         NodeAction listeners $ \sendActions -> do
-            t <- withNode __node
-            action ourVerInfo sendActions `finally` afterNode t
+            bracket (withNode __node) (afterNode) (const (action ourVerInfo sendActions))
```
Comment (dcoutts-2fa) [inline src/Pos/Launcher/Runner.hs]: yes, well spotted

**Comment by avieth** [comment]
> Did you mean to include the receive delay patch in this PR? I assume not so I've not reviewed that patch here.

I needed to bring in a new version of time-warp-nt, one which includes the receive delay feature, so I also brought in the necessary CSL changes. No actual receive delay is used.

File: src/Pos/Launcher/Runner.hs (hunk 2 above)
Comment (avieth) [inline src/Pos/Launcher/Runner.hs]: Done

**Comment by domenkozar** [comment]
@avieth this needs rebasing, otherwise I expect we'll try this early next week.

**Comment by domenkozar** [comment]
@avieth it fails to build (on CI at least):

```
Cabal-simple_mPHDZzAJ_1.24.2.0_ghc-8.0.2: can't find source for
Pos/Statistics/Ekg in src,
.stack-work/dist/x86_64-linux-nix/Cabal-1.24.2.0/build/autogen
```

Can you tell if locally it works for you? Seems like something is missing.

**Comment by deepfire** [comment]
@avieth, the merge conflicts prevent me from working on deploying this.

**Comment by avieth** [comment]
@domenkozar @deepfire It's ready now. The only CI failure is due to out of memory

**Comment by deepfire** [comment]
Current effort is here: https://github.com/input-output-hk/iohk-nixops/tree/devops-74-deploy-ekg

**Comment by domenkozar** [comment]
@deepfire confirmed that we get metrics into datadog using this branch: https://github.com/input-output-hk/iohk-nixops/pull/71

Should we ask @gromakovsky or someone else to review?

**Comment by domenkozar** [comment]
@avieth can you rebase on top of `cardano-sl-0.5-rc` branch and resolve conflicts?

**Comment by avieth** [comment]
> @avieth can you rebase on top of cardano-sl-0.5-rc branch and resolve conflicts?

You sure you want me to rebase here? The merge target is master.

> @avieth I should note that @domenkozar asks if we could do the +RTS -T thing using -with-rtsopts=-T in the .cabal file instead.

Will do.

**Comment by domenkozar** [comment]
We can also make a new PR for `cardano-sl-0.5-rc` but the end result should be the code is in both branches :)

**Comment by dcoutts** [comment]
Note: @arybczak is working on the rebase here (since @avieth is concentrating on other things).

**Comment by domenkozar** [comment]
These have landed in master:

```
$ git branch --contains 382bed1fe5
  cardano-sl-0.5-rc
  cardano-sl-0.5-rc-staging
  devops-100-automate-redemption-cert-gen
* master
```

**Pull request closed without merging**